### ISSUE TITLE
Enigma: Signing option disabled after saving a draft

### ISSUE THREAD

**Issue by githubgrscporg** [roundcube/roundcubemail#6515]
I am using 1.4-beta.  My settings are to digitally sign all emails that I send.  Previous to this version, this has always worked as expected.  With 1.4-beta, the behavior is inconsistent, and I have not been able to (as yet?) determine any pattern.  Sometimes when I send a new email message, it will digitally sign it (and ask for passphrase if required), and other times when sending a new message it simply sends it without signing.  I see similar behavior with messages I forward, and messages to which I am replying.  As far as I can tell, it does not matter to whom I sending the email, or whether there is a single recipient nor group of recipients.

I don't typically encrypt email, but the few times I have elected to encrypt, the behavior has been as I expect -- the message is encrypted.

debug_level is set to 1 in config.inc.php, but I have not seen any reference to enigma or gnupg or any such thing in the logs.

I have tried installing the latest snapshot from github, but have not had luck getting it to work in my system, so I have only my experience with 1.4-beta.  Yes, I have followed the instructions in the INSTALL doc, but to no avail.

I am running Debian Stretch and, with the exception of roundcube, everything else installed on this machine is from the Debian repositories: postgresql, gnupg, openjdk, the various php dependencies, etc.  My mail system includes roundcube, postfix, dovecot, procmail, cyrus sasl, and apache2 is my web server.  This machine is regularly updated.  Oh, it's a vm at linode if that makes any difference.

Please offer any suggestions to help me better report this issue, and my apologies if this has already been addressed.

**Comment by alecpl** [comment]
I have no idea what that could be. Is the signing option checked? Enigma has a debug option, but it will log only gnupg operations no other control info. You say you have php deps in the system. So, you're not using bundled vendor folder? Make sure your deps are in required versions. What GnuPG and PHP version?

**Comment by githubgrscporg** [comment]
The signing option is checked in each compose window ("Digitally sign this message" is enabled), and the following options are enabled in Settings->Encryption:

Enable message encryption and signing
Enable message signatures verification
Enable message decryption
Sign all messages by default

Again, sometimes it signs messages as expected, other times it sends without any attempt to sign.  No pattern to this behavior that I have been able to discern.  It's the oddest behavior.

I am using the new Elastic skin.  I will try in both Larry and Classic as well.

Not sure what is meant by "bundled vendor folder" -- I have been downloading the "Complete" roundcube tarballs (as opposed to "Dependent" or "Framework") for all releases and for this beta if that helps.

Following is the output from 'dpkg-query -l | grep -iE "php|gnupg"'
ii  debian-archive-keyring        2017.5                         all          GnuPG archive keys of the Debian archive
ii  gnupg                         2.1.18-8~deb9u2                amd64        GNU privacy guard - a free PGP replacement
ii  gnupg-agent                   2.1.18-8~deb9u2                amd64        GNU privacy guard - cryptographic agent
ii  libapache2-mod-php7.0         7.0.30-0+deb9u1                amd64        server-side, HTML-embedded scripting language (Apache 2 module)
ii  libassuan0:amd64              2.4.3-2                        amd64        IPC library for the GnuPG components
ii  libgpg-error0:amd64           1.26-2                         amd64        library for common error values and messages in GnuPG components
ii  php-auth-sasl                 1.0.6-3                        all          Abstraction of various SASL mechanism responses
ii  php-common                    1:49                           all          Common files for PHP packages
ii  php-mbstring                  1:7.0+49                       all          MBSTRING module for PHP [default]
ii  php-pear                      1:1.10.1+submodules+notgz-9    all          PEAR Base System
ii  php7.0-cli                    7.0.30-0+deb9u1                amd64        command-line interpreter for the PHP scripting language
ii  php7.0-common                 7.0.30-0+deb9u1                amd64        documentation, examples and common module for PHP
ii  php7.0-gd                     7.0.30-0+deb9u1                amd64        GD module for PHP
ii  php7.0-imap                   7.0.30-0+deb9u1                amd64        IMAP module for PHP
ii  php7.0-json                   7.0.30-0+deb9u1                amd64        JSON module for PHP
ii  php7.0-mbstring               7.0.30-0+deb9u1                amd64        MBSTRING module for PHP
ii  php7.0-opcache                7.0.30-0+deb9u1                amd64        Zend OpCache module for PHP
ii  php7.0-pgsql                  7.0.30-0+deb9u1                amd64        PostgreSQL module for PHP
ii  php7.0-readline               7.0.30-0+deb9u1                amd64        readline module for PHP
ii  php7.0-xml                    7.0.30-0+deb9u1                amd64        DOM, SimpleXML, WDDX, XML, and XSL module for PHP
ii  pinentry-curses               1.0.0-2                        amd64        curses-based PIN or pass-phrase entry dialog for GnuPG

**Comment by alecpl** [comment]
The `/vendor` folder contains all php libs required. So, you're good. Would be good to know if this happens with other skins. What other plugins are you using?

**Comment by githubgrscporg** [comment]
Enigma is the only plugin I am using, or at least it is the only one I have configured.  It seems that filesystem_attachments and jqueryui are also enabled, at least according to the about page.

My apologies, I have sent only a small handful of email since switching to the Larry skin.  No problems so far but, again, I have made only a few tries to test this since switching.

**Comment by githubgrscporg** [comment]
It's been almost three weeks using the Larry skin, and all emails that I have sent have been signed as per my configuration.  I've switched back to the Elastic skin, and will report back.

**Comment by githubgrscporg** [comment]
Went back to Elastic, and first email I sent was not signed.

Any thoughts on how I might collect more data to help with this?

**Comment by alecpl** [comment]
I have an idea what that could be. It should be fixed in git-master. Are you able to test git-master version? If not you have to wait for 1.4-rc (should be released soon).

**Comment by githubgrscporg** [comment]
I've experienced some issues trying to use master, but those are my issues not yours.  I will give it a try.  If I have no success with master, I will test against 1.4-rc when available and will report back here.

**Comment by githubgrscporg** [comment]
I am now using git-master, and my initial impression is that this is no longer an issue.  I am using the Elastic skin.  I'll certainly report back if I see this behavior again.

**Comment by alecpl** [comment]
Is the signing option checkbox in compose window always checked?

**Comment by githubgrscporg** [comment]
Sorry, I didn't notice you had responded.  Yes, the digital signing option is always checked.  To reiterate, this seems only to happen in the Elastic skin.  I have had no issues using the Larry skin (I have not tested this against the classic skin).  I will try a few times accessing from my phone's browser to see if I can replicate from the Elastic/mobile theme.  Again, please let me know if there is anything else I can do to gather logs or other data to help you with this.

**Comment by mail61** [comment]
I have this same issue, 1.4-beta, elastic skin. Seems to be limited to replying to encrypted emails.  You receive an encrypted email, click reply, the "Encrypt this message" checkbox is checked already, send your reply, it will be sent un-encrypted.  If during the reply process you uncheck and then recheck "Encrypt this message" it will send encrypted however.  Hopefully this is a minor issue that can be addressed quickly as my users think they are sending encrypted replies but they are not.

**Comment by mail61** [comment]
Is there anything I can contribute further? The status is "needing feedback".

**Comment by alecpl** [comment]
@mail61 your issue is fixed in git-master. The issue here is about signing.

I was now able to reproduce the issue. Investigating.

**Comment by alecpl** [comment]
Happened to me once, but now I can't reproduce. We'll need to find some pattern.

**Comment by alecpl** [comment]
I think I finally fixed this in b24869b0cd52.

**Comment by githubgrscporg** [comment]
This still happens, though with less frequency than before.  Since getting a new copy from git on June 12, I have seen this behavior when replying to some, but not all, unsigned emails.  The only signed emails I have received in this time are test emails from myself, and replies to those seem to all be signed as well.  New emails I have sent have also been signed.  Again, I think the only time I have seen this behavior lately is when replying to unsigned email.  I'm sorry to have to report this.

**Comment by githubgrscporg** [comment]
Oh, I have been using the Elastic skin exclusively since I pulled from git on June 12.

**Comment by githubgrscporg** [comment]
This happened just now sending a new email.  This is the first time it has happened, I think, on new email since I have installed git-master on June 12 -- since then, and until today, the only time email has been sent unsigned was when replying to some-but-not-all unsigned emails.  "Digitally sign this message" is selected in my default compose mail settings, and was selected on this email I just sent.  Sorry to have to bother you with this... I know you have closed the ticket.

**Comment by githubgrscporg** [comment]
Similar to last report, I am using only the Elastic skin at this time.

**Comment by githubgrscporg** [comment]
I am using the Elastic skin on 1.4rc2, and the same behavior exists.  Most of the time, a message is signed, but occasionally a message is sent without being signed.  I still have no discernible pattern.  This happens only in the Elastic skin.  As this report is closed, but the behavior persists, should I open a new ticket?

**Comment by alecpl** [comment]
What web browser are you using? What other plugins are you using?

**Comment by githubgrscporg** [comment]
Shoot, I didn't think of documenting this before.  Firefox, regularly updated: v 69.0.1 at this time.  OS is (usually) Debian 10 (Buster) Linux, but was Debian 9 (Stretch?) when I first reported this (I'm using Firefox as downloaded from mozilla.org rather than the Firefox-ESR which is included in most Debian desktops).  I also use Windows 10 from time to time also with frequently updated Firefox.  Oh, at work my workstation is Ubuntu 18.04 LTS also with the downloaded and updated Firefox; it was Ubuntu 16.04 LTS when I first reported this.

The only installed extension is Duck Duck Go Privacy Essentials, and the only plugin I see is OpenH264 Video Codec by Cisco.  I used to also have the Video DownloadHelper (?) and HTML Validator extensions installed, but I removed them some time ago.  I have experienced this behavior both with and without DownloadHelper and HTML Validator extensions.  Not sure about Duck Duck Go.

I know I have seen this behavior at work and also at home on my Linux machines, and I *think* I have also seen it on Windows, but cannot confirm this.  I do not recall running Chrome or Chromium much in the last year or so, but I will try using those as well.

**Comment by alecpl** [comment]
Found it! This is caused by draft auto-save. When it runs at least once before you press Send it unsets the sign-flag state. A workaround is to, before you press Send, uncheck and check the checkbox again. It should reset the state to a proper value.

 I'll work on a fix.

**Comment by githubgrscporg** [comment]
Great to see progress on this!

As for browser-type: I spent the last three weeks using Chrome, and did not experience this issue.  Went back to Firefox a few days ago, and have seen it once.  I have sent fifteen emails from Chrome and two from Firefox in that time.  Not sure if that is significant or not, but I figured to report back since you had asked.

Anyway, I do have the "Automatically save draft" setting set to "every 5 minute(s)".  It never occurred to me to time how long I spent drafting an email, nor whether or not a draft had been saved before I sent.  I will put this to the test to see if I can confirm this behavior.
